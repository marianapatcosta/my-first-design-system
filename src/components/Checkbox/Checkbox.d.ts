import React, { ChangeEvent, CSSProperties } from 'react';
export interface CheckboxProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * Checkbox's label
     */
    label: string;
    /**
     * Checkbox's name
     */
    name?: string;
    /**
     * Is Checkbox checked?
     */
    checked: boolean;
    /**
     * Is Checkbox disabled?
     */
    disabled?: boolean;
    /**
     * Change handler
     */
    onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}
export declare const Checkbox: React.FC<CheckboxProps>;
