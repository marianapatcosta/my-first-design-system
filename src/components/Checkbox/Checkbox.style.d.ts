export declare const StyledCheckbox: import("styled-components").StyledComponent<"label", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledCheckboxToggleTick: import("styled-components").StyledComponent<"img", any, {}, never>;
export declare const StyledCheckboxToggle: import("styled-components").StyledComponent<"div", any, {
    disabled?: boolean | undefined;
}, never>;
export declare const StyledCheckboxInput: import("styled-components").StyledComponent<"input", any, {}, never>;
