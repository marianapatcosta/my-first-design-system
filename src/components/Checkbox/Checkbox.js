var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Tick } from '../../assets/icons';
import { StyledCheckbox, StyledCheckboxToggle, StyledCheckboxInput, StyledCheckboxToggleTick, } from './Checkbox.style';
export var Checkbox = function (_a) {
    var checked = _a.checked, name = _a.name, disabled = _a.disabled, label = _a.label, className = _a.className, onChange = _a.onChange, props = __rest(_a, ["checked", "name", "disabled", "label", "className", "onChange"]);
    return (_jsxs(StyledCheckbox, __assign({ className: className, disabled: disabled }, props, { children: [_jsx(StyledCheckboxInput, { type: 'checkbox', name: name || label, checked: checked, disabled: disabled, onChange: onChange }, void 0),
            _jsx(StyledCheckboxToggle, __assign({ disabled: disabled }, { children: checked && _jsx(StyledCheckboxToggleTick, { alt: 'checkbox-tick', src: Tick }, void 0) }), void 0), label] }), void 0));
};
Checkbox.defaultProps = {
    id: '',
    className: '',
    style: {},
    name: '',
    label: '',
    checked: false,
    disabled: false,
    onChange: function () { return null; },
};
