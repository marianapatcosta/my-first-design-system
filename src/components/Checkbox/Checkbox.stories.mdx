import { Meta, Story, ArgsTable, Canvas, Preview } from '@storybook/addon-docs/blocks';
import { Checkbox } from "./Checkbox";

<Meta title="Design System/Checkbox" component={Checkbox} />

# Checkbox
<br/><br/> 

### Allows to alternate between two values

Use checkbox when the value alternates between true and false

<Preview withToolbar>
  <Story name="Standard">
     <Checkbox checked={false} label="Checkbox" />
  </Story>
</Preview>

<ArgsTable of={Checkbox} />

### Playground

Experiment with Checkbox Component with the Controls addon in the Canvas tab.

export const Template = (args) => <Checkbox {...args} />

<Canvas>
  <Story name="Controls" args={{
 label: 'Checkbox',
  checked: false,
  }}>
    {Template.bind({})}
  </Story>
</Canvas> 