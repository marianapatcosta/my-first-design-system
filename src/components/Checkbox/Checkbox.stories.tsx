import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Checkbox, CheckboxProps } from './Checkbox';

export default {
  title: 'Checkbox',
  component: Checkbox,
} as Meta;

const Template: Story<CheckboxProps> = (args) => <Checkbox {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Checkbox',
  checked: false,
};
