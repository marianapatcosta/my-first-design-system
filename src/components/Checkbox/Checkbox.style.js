var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, disabledStyles, colors, truncateStyles, } from '../../styles';
export var StyledCheckbox = styled.label(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", ";\n  ", ";\n  display: flex;\n  width: 100%;\n  font-size: 95%;\n  color: ", ";\n\n  ", "\n"], ["\n  ", ";\n  ", ";\n  display: flex;\n  width: 100%;\n  font-size: 95%;\n  color: ", ";\n\n  ", "\n"])), baseStyles, truncateStyles, colors.highlight, function (_a) {
    var disabled = _a.disabled;
    return disabled && "color: " + colors.disabled + ";";
});
export var StyledCheckboxToggleTick = styled.img(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  width: 1rem;\n  height: 1rem;\n  ", ";\n"], ["\n  width: 1rem;\n  height: 1rem;\n  ", ";\n"])), colors.iconHighlight);
export var StyledCheckboxToggle = styled.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  position: relative;\n  width: 1rem;\n  height: 1rem;\n  margin: 0.125rem;\n  margin-right: 0.5rem;\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  border-radius: 0.2rem;\n  background-color: ", ";\n  transition: background-color 0.5s ease-in-out;\n  cursor: pointer;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  ", "\n"], ["\n  display: flex;\n  align-items: center;\n  position: relative;\n  width: 1rem;\n  height: 1rem;\n  margin: 0.125rem;\n  margin-right: 0.5rem;\n  box-shadow: 0 0.063rem 0.25rem ", ";\n  border-radius: 0.2rem;\n  background-color: ", ";\n  transition: background-color 0.5s ease-in-out;\n  cursor: pointer;\n\n  :hover {\n    opacity: 0.75;\n  }\n\n  ",
    "\n"])), colors.shadow, colors.primary, function (_a) {
    var disabled = _a.disabled;
    return disabled &&
        "\n      background-color: " + colors.disabled + ";\n      " + disabledStyles + "\n    ";
});
export var StyledCheckboxInput = styled.input(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  position: absolute;\n  opacity: 0;\n  z-index: -1;\n\n  :checked + ", " {\n    background-color: ", ";\n  }\n\n  // when focus, adds a border to its sibling StyledCheckboxToggle\n  :focus-visible + ", " {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) + ", " {\n    outline: none;\n  }\n"], ["\n  position: absolute;\n  opacity: 0;\n  z-index: -1;\n\n  :checked + ", " {\n    background-color: ", ";\n  }\n\n  // when focus, adds a border to its sibling StyledCheckboxToggle\n  :focus-visible + ", " {\n    outline: 0.125rem solid ", ";\n  }\n\n  :focus:not(:focus-visible) + ", " {\n    outline: none;\n  }\n"])), StyledCheckboxToggle, colors.highlight, StyledCheckboxToggle, colors.highlight, StyledCheckboxToggle);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
