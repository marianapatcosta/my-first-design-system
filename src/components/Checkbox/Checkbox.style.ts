import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';

export const StyledCheckbox = styled.label<{ disabled?: boolean }>`
  ${baseStyles}
  display: flex;
  width: 100%;
  font-size: 95%;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: ${colors.highlight};

  ${({ disabled }) =>
    disabled
      ? `
      color: ${disabled};
    `
      : ''}
`;

export const StyledCheckboxToggleTick = styled.img`
  width: 1rem;
  height: 1rem;
  filter: invert(98%) sepia(26%) saturate(2%) hue-rotate(98deg) brightness(116%)
    contrast(100%);
`;

export const StyledCheckboxToggle = styled.div<{ disabled?: boolean }>`
  display: flex;
  align-items: center;
  position: relative;
  width: 1rem;
  height: 1rem;
  margin: 0.125rem;
  margin-right: 0.5rem;
  box-shadow: 0 0.063rem 0.25rem ${colors.shadow};
  border-radius: 0.2rem;
  background-color: ${colors.primary};
  transition: all 0.5s ease-in-out;
  cursor: pointer;

  :hover {
    opacity: 0.75;
  }

  ${({ disabled }) =>
    disabled
      && `
      background-color: ${colors.disabled};
      opacity: 0.3;
      pointer-events: none;
      cursor: default;
    `}
`;

export const StyledCheckboxInput = styled.input`
  position: absolute;
  opacity: 0;
  z-index: -1;

  :checked + ${StyledCheckboxToggle} {
    background-color: ${colors.highlight};
  }

  // when focus, adds a border to its sibling StyledCheckboxToggle
  :focus + ${StyledCheckboxToggle} {
    outline: 0.125rem solid ${colors.highlight};
  }

  :focus:not(:focus-visible) + ${StyledCheckboxToggle} {
    outline: none;
  }
`;
