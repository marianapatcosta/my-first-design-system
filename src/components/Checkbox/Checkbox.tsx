import React, { ChangeEvent, CSSProperties } from 'react';
import { Tick } from '../../assets/icons';
import {
  StyledCheckbox,
  StyledCheckboxToggle,
  StyledCheckboxInput,
  StyledCheckboxToggleTick,
} from './Checkbox.style';

export interface CheckboxProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * Checkbox's label
   */
  label: string;
  /**
   * Checkbox's name
   */
  name?: string;
  /**
   * Is Checkbox checked?
   */
  checked: boolean;
  /**
   * Is Checkbox disabled?
   */
  disabled?: boolean;
  /**
   * Change handler
   */
  onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}

export const Checkbox: React.FC<CheckboxProps> = ({
  checked,
  name,
  disabled,
  label,
  className,
  onChange,
  ...props
}) => {
  return (
    <StyledCheckbox className={className} disabled={disabled} {...props}>
      <StyledCheckboxInput
        type='checkbox'
        name={name || label}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
      />
      <StyledCheckboxToggle disabled={disabled}>
        {checked && <StyledCheckboxToggleTick alt='checkbox tick' src={Tick} />}
      </StyledCheckboxToggle>
      {label}
    </StyledCheckbox>
  );
};

Checkbox.defaultProps = {
  id: '',
  className: '',
  style: {},
  name: '',
  label: '',
  checked: false,
  disabled: false,
  onChange: () => null,
};
