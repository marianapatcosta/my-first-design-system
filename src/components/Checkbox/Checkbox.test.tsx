import React from 'react';
import { render, getByText, fireEvent } from '@testing-library/react';
import { Checkbox } from './Checkbox';

describe('Checkbox', () => {
  const defaultProps = {
    label: 'Checkbox',
    checked: false,
    onChange: () => null,
  };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Checkbox {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display label', () => {
    const { container } = render(<Checkbox {...defaultProps} />);

    getByText(container, 'Checkbox');
  });

  test('should handle change events', () => {
    const onChangeMock = jest.fn();
    const { getByLabelText } = render(
      <Checkbox {...defaultProps} onChange={onChangeMock} />
    );
    const checkbox = getByLabelText(/Checkbox/i) as HTMLInputElement;
    fireEvent.click(checkbox);

    expect(onChangeMock).toBeCalled();
    /* expect(checkbox).not.toBeChecked();
    expect(checkbox.checked).toBeTruthy();
    fireEvent.click(checkbox)
    expect(checkbox.checked).toBeFalsy(); */
  });

  test('should return null on change if onChange prop is not passed', () => {
  const result = Checkbox.defaultProps?.onChange && Checkbox.defaultProps.onChange();
    expect(result).toBeNull();
  });

  describe('Disabled Checkbox', () => {
    test('should display label', () => {
      const { container } = render(
        <Checkbox {...defaultProps} disabled={true} />
      );

      getByText(container, 'Checkbox');
    });

    test('should be disabled', () => {
      const {
        container: { firstChild },
      } = render(<Checkbox {...defaultProps} disabled />);

      expect(firstChild).toHaveAttribute('disabled');
    });

    test('should not handle change events', () => {
      const onChangeMock = jest.fn();
      const {
        container: { firstChild },
      } = render(
        <Checkbox {...defaultProps} disabled={true} onChange={onChangeMock} />
      );

      firstChild && fireEvent.click(firstChild);

      expect(onChangeMock).not.toBeCalled();
    });
  });
});
