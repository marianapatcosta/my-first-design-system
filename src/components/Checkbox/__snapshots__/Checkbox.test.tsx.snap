// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders without errors and matches snapshot 1`] = `
<label
  class="sc-bdfBwQ cDBEUr"
  id=""
>
  <input
    class="sc-hKgILt fEjjZ"
    name="Checkbox"
    type="checkbox"
  />
  <div
    class="sc-dlfnbm cVmCWh"
  />
  Checkbox
</label>
`;
